rest_framework_api_key/models.py

```diff
@@ -38,7 +39,9 @@ def is_valid(self, key: str) -> bool:
 class APIKey(models.Model):
     objects = APIKeyManager()
 
-    id = models.CharField(max_length=100, unique=True, primary_key=True)
+    prefix = models.CharField(max_length=8, unique=True)
+    hashed_key = models.CharField(max_length=32)
```

Comment:
Ah sorry about that. I took the 32 character limit from the existing code but of course, in retrospect, that was the **secret key** length rather than the hashed key. I'll update that immediately.

I'll change the AlterField to the a RemoveField and AddField too.

With respect to the reverse migration, it would be nice to have but I can't see that it would ever actually be needed. 

We should also mark that this PR is a potentially breaking change especially if someone has used the keys in this model as a foreign key in another model. I can't see why anyone would but there is always the chance.

Reply:
> if someone has used the keys in this model as a foreign key in another model

e.g. something like this I suppose?

```python
class Person(models.Model):
    api_key = models.ForeignKey("APIKey")
```

I think some people are indeed using this pattern (see #20 for example), so we might need to see if there's a way to make this migration compatible with foreign keys.

My guess is that this would require to update the FK on the dependant models (i.e. making `api_key_id` point to the new integer value), but I don't know how easy this is to do with Django.

Also this is something to be done on our user's side, so we'll probably want to add some upgrade instructions.

Reply:
I'm not able to test it at the moment but the issue would be the changing of the underlying field type from char to int. So I would imagine the steps to migrate any foreign keys would be something along the lines of,

1. Before upgrading the package, change the foreign key to a regular char field and run that migration so that the records are not deleted during the upgrade

1. Upgrade the package and run the package migrations

1. Create a new nullable foreign key field to the APIKey model and add a data migration to correctly find the originally linked records by their prefix and hash.

1. Update the foreign key to remove the nullable attribute

Reply:
Okay, thanks!

Tbh, that sounds like a huge pain though. As a downstream user, I wouldn’t want to have to go through all that. We risk creating an upgrade wall, and plain frustration.

I remember we mentioned using the prefix as a PK. I was initially against it because it is tightly related to the cryptographic scheme. But at the same time, I think this scheme is quite mature now. It works well, and I don’t see it changing in the future, at least not in v1.

Since the prefix is a string too, we wouldn’t have the issue of type conversion, and we’d be able to use AlterField, thereby preventing any cascading as the column wouldn’t need to be dropped.

My only concern is unicity of the prefix: it is enforced at the DB level, but it’s still generated randomly without consideration of existing prefixes for now. This can be fixed, I believe.

So, what do you think? Would that be a better idea?